components/higher-order/with-lazy-dependencies/index.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+const alreadyLoaded = {};
+
+function loadScript( url ) {
+	if ( alreadyLoaded[ url ] ) {
+		return Promise.resolve();
+	}
+
+	return new Promise( ( resolve, reject ) => {
+		const script = document.createElement( 'script' );
+		script.src = url;
+		script.onload = () => {
+			alreadyLoaded[ url ] = true;
+			resolve();
+		};
+		script.onerror = () => reject();
+
+		document.head.appendChild( script );
+	} );
+}
+
+function loadStyle( url ) {
+	if ( alreadyLoaded[ url ] ) {
+		return Promise.resolve();
+	}
+
+	return new Promise( ( resolve, reject ) => {
+		const style = document.createElement( 'link' );
+		style.rel = 'stylesheet';
+		style.href = url;
+		style.onload = () => {
+			alreadyLoaded[ url ] = true;
+			resolve();
+		};
+		style.onerror = () => reject();
+
+		document.head.appendChild( style );
+	} );
+}
+
+function loadDependencies( scripts, styles ) {
+	const promises = [];
+
+	if ( scripts && scripts.length > 0 ) {
+		const url = `/wp-admin/load-scripts.php?load=${ scripts.join( ',' ) }`;
+		promises.push( loadScript( url ) );
+	}
+
+	if ( styles && styles.length > 0 ) {
+		const url = `/wp-admin/load-styles.php?load=${ styles.join( ',' ) }`;
+		promises.push( loadStyle( url ) );
+	}
+
+	return Promise.all( promises );
+}
+
+export default ( { scripts, styles } ) => WrappedComponent =>
+	class LazyDependencyComponent extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.state = {
+				hasLoaded: false,
+			};
+		}
+
+		componentDidMount() {
+			loadDependencies( scripts, styles ).then( () => {
+				this.setState( { hasLoaded: true } );
+			} );
```

Comment:
What if any of the dependencies won't load?

Reply:
Aye, we'll need to add a loading state and error state to `withLazyDependencies`.

I'm thinking something like this:

```js
const LazyComponent = withLazyDependencies( {
    scripts: [ 'foo', 'bar' ],
    styles: [ 'baz' ],
    loadingComponent() {
        return <Spinner />;
    },
    errorComponent( message ) {
        return <div>{ sprintf( __( 'Uh oh! %s' ), message ) }</div>
    },
} )( MyComponent );
```